docs/architecture/adr-019-protobuf-state-encoding.md

```diff
@@ -0,0 +1,245 @@
+# ADR 019: Protocol Buffer State Encoding
+
+## Changelog
+
+- 2020 Feb 15: Initial Draft
+
+## Status
+
+Accepted
+
+## Context
+
+Currently, the Cosmos SDK utilizes [go-amino](https://github.com/tendermint/go-amino/) for object encoding
+over the wire bringing parity between logical objects and persistence objects. From the Amino docs:
+
+> Amino is an object encoding specification. It is a subset of Proto3 with an extension for interface
+> support. See the [Proto3 spec](https://developers.google.com/protocol-buffers/docs/proto3) for more
+> information on Proto3, which Amino is largely compatible with (but not with Proto2).
+>
+> The goal of the Amino encoding protocol is to bring parity into logic objects and persistence objects.
+
+Amino also aims to have the following goals (not a complete list):
+
+- Binary bytes must be decode-able with a schema.
+- Schema must be upgradeable.
+- The encoder and decoder logic must be reasonably simple.
+
+However, we believe that Amino does not fulfill these goals completely and does not fully meet the
+needs of a truly flexible cross-language and multi-client compatible encoding protocol in the Cosmos SDK.
+Namely, Amino has proven to be a big pain-point in regards to supporting object serialization across
+clients written in various languages while providing virtually little in the way of true backwards
+compatibility and upgradeability. Furthermore, through profiling and various benchmarks, Amino has
+been shown to be an extremely large performance bottleneck in the Cosmos SDK <sup>1</sup>. This is
+largely reflected in the performance of simulations and application transaction throughput.
+
+Thus, we need to adopt an encoding protocol that meets the following criteria:
+
+- Language agnostic
+- Platform agnostic
+- Rich client support and thriving ecosystem
+- High performance
+- Minimal encoded message size
+- Codegen-based over reflection-based
+- Supports backward and forward compatibility
+
+## Decision
+
+We will adopt [Protocol Buffers](https://developers.google.com/protocol-buffers) for serializing
+persisted structured data in the Cosmos SDK while providing a clean mechanism and developer UX for
+applications wishing to continue to use Amino. We will provide this mechanism by updating modules to
+accept a codec interface, `Marshaler`, instead of a concrete Amino codec. Furthermore, the Cosmos SDK
+will provide three concrete implementations of the `Marshaler` interface: `AminoCodec`, `ProtoCodec`,
+and `HybridCodec`.
+
+The `AminoCodec` will use Amino for both binary and JSON encoding. The `ProtoCodec` will use Protobuf
+for or both binary and JSON encoding. Finally, `HybridCodec` will use Amino for JSON encoding and
+Protobuf for binary encoding. A module will typically use `HybridCodec` as the concrete codec it
```

Comment:
It'd be great to also explain why Amino was chosen for JSON encoding instead of Protobuf, i.e why do we need an `HybridCodec`

Reply:
It’s probably worth noting both the type confusion mitigations and canonicalization in amino JSON

Reply:
That would be nice, or a link somewhere if it is documented what is the difference between both json (I currently don't know)

Reply:
Added more context on the rationale and ultimate goal.